napari/_qt/widgets/magic_gui_base_widget.py

```diff
@@ -0,0 +1,60 @@
+from qtpy.QtCore import QObject, Signal
+
+
+class BaseMagicSetting(QObject):
+    """Base class that helps use Magic GdUI widgets in the json schema widget
+    buider.
+    """
+
+    valueChanged = Signal(dict)
+
+    def get_mgui(self):
+        """Get magic gui widget."""
+        raise NotImplementedError()
+
+    def __init__(self, description=None):
+        super().__init__()
+        self._widget = self.get_mgui()
+        self._description = description
+
+        @self._widget.changed.connect
+        def _call_magic_gui(value: dict = {}):
+            self.valueChanged.emit(value())
+
+    def __getattr__(self, attribute):
+        """Method that will retrieve needed information
+        from the actual widget if not found in the class itself.
+        """
+        try:
+            return self.__getattribute__(attribute)
+        except AttributeError:
+            try:
+                return getattr(self._widget.native, attribute)
+            except AttributeError as e:
+                raise (e)
+
+    def setDescription(self, value):
+        """Set description of magic gui widget.
+
+        Parameters
+        ----------
+        value: str
+            Description for widget.
+        """
+        self._description = value
+
+    def setToolTip(self, value):
+        """Set tooltip for each widget in the magic gui widget.
+
+        Note: Currently, there is only 1 tooltip that will be used
+        for all widgets.
+
+        Parameters
+        ----------
+        value: str
+            Tooltip for widget.
+        """
+
+        if value:
+            for widget in self._widget._list:
+                widget.tooltip = value
```

Comment:
Still, we use here magicgui private API. It is a really bad idea. If it is not possible to be done with public API then better not do this or ask @tlambert03 for a hint.

Reply:
Get_mgui() is abstract here and doesn't have a type annotation, so I can't tell what `self._widget` actually is.  Not all magicgui widgets have a `_list` attribute, so that's also bad.

But, if it's a Container, just iterate over self._widget directly.  (But I would definitely clarify here what get_mgui is and isn't allowed to return)